## Researchproject: 
### Comparison of hybrid/longread assembly software and elaboration of a best practice
***
***
### Overview

With the advance of new sequencing technologies, the demands on and possibilities of bioinformatic software for data analysis and genome reconstruction are increasing. Accordingly, a large variety of tools for genome reconstruction has been established over the years.<sup>1</sup>
This research project tries to give an overview of the state of the art software for whole genome assembly of prokaryotes: Based on a literature review a promising pre-selection of assembly software is made and it is investigated which metrics are useful to assess assembly quality. In a subsequent analysis the following questions are dealt with in particular:

- Do some assemblers perfrom significantly better than others?
- What edge cases (regarding read data and the underlying genome) are problematic for which assemblers?
- Can longread only assemblies compete with hybrid assemblies?
- How low can read coverage become until a significant drop in assembly quality is observed? (This also aims to reduce costs for sequencing experiments.)

The final goal of the project is the elaboration of a best practice (possibly comprising a script) based on the findings of the assembly software evaluation. The evaluation covers different project phases which are discussed in more detail below. To keep track of the projects progress the different phases are marked with a checkmar once completed.

- ##### [I. Literature Research](user-content-i.-literature-research) :heavy_check_mark:
- ##### [II. First round of evaluation](user-content-ii.-first-round-of-evaluation) :heavy_check_mark:
- ##### [III. Acquisition of more complex input data](user-content-iii.-acquisition-of-more-complex-input-data)
- ##### [IV. Second round of evaluation](user-content-iv.-second-round-of-evaluation)
- ##### [V. Conclusion and best practice](user-content-v.-conclusion-and-best-practice)
- ##### [Data](user-content-data)
- ##### [References](user-content-references)
&nbsp;
***
### I. Literature research

...

### Results
...

***
### II. First round of evaluation
...

### Results
...
***
### III. Acquisition of more complex input data
...

### Results
...
***
### IV. Second round of evaluation
...

### Results
...
***
### V. Conclusion and best practice
...

### Results
...
***
### Data
...

***
### References
<sup>1</sup> [Alice Maria Giani, et al., *Long walk to genomics: History and current approaches to genome sequencing and assembly*, Computational and Structural Biotechnology Journal, Volume 18, 2020, Pages 9-19](https://www.sciencedirect.com/science/article/pii/S2001037019303277)
